### DISCUSSION TITLE
Primary Constructors and private fields naming convention

### DISCUSSION THREAD

**Discussion by philfontaine** [dotnet/csharplang#7565] [Q&A] [answered]
The [C# style guide](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names#camel-case) recommends prefixing internal and private variables with `_`.

> Use camel casing ("camelCasing") when naming private or internal fields and prefix them with _.

> When editing C# code that follows these naming conventions in an IDE that supports statement completion, typing _ will show all of the object-scoped members.

With primary constructors, this advice becomes outdated or problematic since the primary constructor arguments are not prefixed

We will either:
1. Have a mix of prefixed and non-prefixed private fields
2. Have non-prefixed private fields only
3. Prefix the primary constructor arguments (probably not).

My question is, will that style guide be updated to suggest a way forward? Is there already a consensus?

**Comment by HaloFour** [comment, marked as answer]
Primary constructor parameters aren't fields as far as the C# language is concerned.  They are captures, and the actual mechanism used to capture the value of the parameter beyond the lifecycle of the constructor (if necessary) is an implementation detail.  So I think the naming guideline would be to treat them as normal parameters, as you would do with method parameters that are captured by lambdas.

**Reply by CyrusNajmabadi** [reply]
@HaloFour is correct.  These are not fields.  The rules about parameters apply to them and they should be named as such.  

For records, as they are actually member declarations (properties), they should be named as properties).

**Reply by insouciiance** [reply]
The problem to me is, even though primary parameters are not private fields, when I see camelCase in a method body for example, I immediately treat the variable as a method parameter. And the fact that there is no differentiation between naming just adds cognitive load to figure out where it comes from.

**Reply by philfontaine** [reply]
Yes, I have the same opinion.

I get that from a style guide standpoint there is no ambiguity, but in reality the DX is getting worse for reasons you just explained.

**Reply by CyrusNajmabadi** [reply]
> when I see camelCase in a method body for example, I immediately treat the variable as a method parameter

It is a method parameter (constructor are just methods).  :)

> And the fact that there is no differentiation between naming just adds cognitive load to figure out where it comes from.

It comes from the parameter list.  We could consider a way in the IDE to let you give a different classification for them.  You can also choose to name them however you want if you don't like the chosen naming convention.  Naming conventions are just conventions.  If they're not good for your use case, feel free to pick anything else :)

**Reply by insouciiance** [reply]
Yes, IDE highlighting would def be helpful!

To me, it's either that, or appending `_` to primary parameter names where they would otherwise be private fields

**Reply by CyrusNajmabadi** [reply]
> where they would otherwise be private fields

To be clear.  They are never private fields.  They are constructor parameters :)

Compiler is free to implement taht however it wants.  For example, on a runtime where parameters can just be referenced from any nested code, they would remain parameters.  Or they could be associated in a side-table associated with teh instance.  *Or* a compiler could put them in fields, etc.  But all of that is 100% an impl detail, and is not exposed from the language level (and can change at any time).

A good example, is that the compiler can be smart and elide all storage *altogether* when possible.  As a field, this wouldn't be possible, as it would be part of the contract of the instance.  But as a parameter, this is very possible.

**Reply by DavidArno** [reply]
> The problem to me is, even though primary parameters are not private fields, when I see camelCase in a method body for example, I immediately treat the variable as a method parameter. And the fact that there is no differentiation between naming just adds cognitive load to figure out where it comes from.

@insouciiance,

One solution to that is to declare your own private field and assign the parameter to it and then only reference the field. That way, you can stick to your current naming convention. 

As a bonus, since the team have chosen not to provide a means of marking those PC parameters as readonly, you can make your own fields readonly and forbid access to the PC parameters outside of construction/initialisation via an analyzer. I can well imagine this approach quickly becoming the consensus best practice approach as many developers will find the "*they aren't fields, they are deliberately mutable captures*" line to be too "hand-wavy" for their liking.

**Reply by rcdailey** [reply]
I just upgraded to .NET 8 and started playing with PC for my classes and I quickly had the same concern about naming. I think the technical explanation of how they are not fields is informative, but sort of misses the point, at least from my perspective. I think the value of the underscore prefix is not to identify fields, but to identify scope. Something named `_foo` can be immediately identified as something available in class scope, and that isn't exclusively within the scope of the method I'm looking at. That "something" might be a field, or some impl detail.

Is that really valuable as far as PC arguments are concerned, though? I'm not sure. For example, if I'm reviewing code in a method, does it really matter to me if `foo` is from a PC or method arguments? I think now that .NET 8 / C# 12 is official, I'll need to get through the honeymoon phase to settle my feelings.

Either way, I appreciate the explanations here.

**Reply by MarkJacksonRG** [reply]
I have the same concern as @rcdailey . IMHO the underscore convention is to identify the _scope_ of the identifier. It's very important to know the scope. It's less important to know whether it's a capture or a field.

I get that primary constructor parameters *are* parameters. I feel a better person for knowing that. But the captures can be used *in different methods* where they *aren't* parameters. So it is not the whole story to say they are "just" parameters. If they were "just" parameters, they would be useless! 

I feel the principle of least surprise is relevant. To me, names without underscores do look natural in the primary constructor, but unnatural in the rest of the class body. The primary constructor is short, simple and unlikely to cause issues. I spend far more time reading the rest of the class, which is longer, more complex, and more likely to cause issues. So it's more important for it to be understandable. 

I notice the C# style guide has this tip below. It seems misleading now. Typing _ doesn't show the PC captures. It's a symptom of this bigger problem.
>When editing C# code that follows these naming conventions in an IDE that supports statement completion, typing _ will show all of the object-scoped members.

Personally I would prefer one of these options 
 - add the underscore prefix to the primary constructor parameter 
 - drop the guidance that private fields should have underscore prefix. I believe Java doesn't have that convention and they seem to manage. IDEs could color the primary constructor captures differently: [Rider does](https://blog.jetbrains.com/dotnet/2023/11/23/primary-constructors-using-csharp-12-in-rider-and-resharper/#semantic-highlighting), don't know about others.

**Reply by yaakov-h** [reply]
on a technicality, primary constructor parameters aren't "object-scoped members" because they aren't members, they are captured variables (like in a lambda).

**Reply by DavidArno** [reply]
@yaakov-h,

Could you give an example of how they could be captured without being members? "_That they are captured as fields is an implementation detail_" currently seems a meaningless statement as it seems the only possible implementation option available without descending into the realms of silliness. The reality is, they are members (though for "reasons", the fields are given unspeakable names and the parameters are mapped to them by the tooling). It feels like either I'm missing something here or this feature has descended into the realm of full-blown "emperor's new clothes".

**Reply by CyrusNajmabadi** [reply]
> Could you give an example of how they could be captured without being members?

A trivial and reasonable approach would be too use a ConditionalWeakTable to store them. This is what roslyn does in some cases, especially to ensure individual object sizes are small, especially for uncommonly accessed data.

**Reply by hakenr** [reply]
Any idea how to set the underscore prefix for primary constructor parameters  as a `.editoconfig` rule?

**Reply by HeinziAT** [reply]
When choosing a convention for your PC parameter names, consider the fact that, unlike private fields, PC parameter names are part of the public contract of your class (think about a client calling your constructor with named arguments). Thus, (a) changing their name is technically a breaking change and (b) starting publicly visible identifiers with underscores feels non-idiomatic in C#.

**Reply by DavidArno** [reply]
@hakenr,

Whilst I don't agree with @HeinziAT regarding this approach resulting in non-idiomatic C# (that went out the window with records and the fact that so much tooling now insists on PascalCase names for record PC parameters - another "I told you so" argument that I lost with the language team), it's worth bearing in mind that other solutions exist here that might better satisfy your needs. For example, [@FaustVX has produced a source generator that will create `_foo` fields for `foo` parameters for you](https://github.com/dotnet/csharplang/discussions/7667#discussioncomment-7634168).

**Reply by baterja** [reply]
I found an MS page that states that fields are generated for PC parameters there: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/instance-constructors#primary-constructors

>  When a primary constructor parameter is used, the compiler captures the constructor parameter in a private field with a compiler-generated name. (...) If the type includes the record modifier, the compiler instead synthesizes a public property with the same name as the primary constructor parameter.

I also find it confusing that primary constructors work differently for records than for (typical) classes.

**Reply by CyrusNajmabadi** [reply]
@baterja that page is describing an implementation detail.  Similar to how you can find descriptions about how lambdas are compiled.  But it's just an impl detail, nothing more.

**Reply by baterja** [reply]
It's not more of a technical detail [than any other behavior accessible with reflection](https://www.hyrumslaw.com/).

**Reply by HaloFour** [reply]
It's still an implementation detail, and not one that one should depend upon as the compiler can change what it does.  The language specification only defines that the parameter is in scope for the body of the class, but does not specify how.
